SIMD values should be saved during OSR entry.

justinmichaud opened the pull request:
#### a820d89ad9983a5a3430269d43b048f0026eb855
<pre>
SIMD values should be saved during OSR entry.
<a href="https://bugs.webkit.org/show_bug.cgi?id=249509">https://bugs.webkit.org/show_bug.cgi?id=249509</a>

Reviewed by Yusuke Suzuki.

We update the masm probe, b3 and air generators to have a separate
osr entry path that saves the full value of vector registers.

A CPUState can now either have packed fprs or packed vectors, depending on
if the context&apos;s Air::Code has usesSIMD == true. This ensures that we
can still load/store pair in the common case, but we do waste some space.

The entry scratch buffer size is doubled when SIMD is used. While this is
wasteful, it is less likely to have bugs. Since this will only regress SIMD
functions, it is hopefully not too bad.

* JSTests/wasm/stress/simd-osr.js:
* Source/JavaScriptCore/assembler/MacroAssembler.h:
* Source/JavaScriptCore/assembler/MacroAssemblerARM64.cpp:
(JSC::MacroAssembler::probe):
* Source/JavaScriptCore/assembler/ProbeContext.h:
(JSC::Probe::CPUState::fpr):
(JSC::Probe::CPUState::vector):
(JSC::Probe::Context::vector):
* Source/JavaScriptCore/wasm/WasmAirIRGeneratorBase.h:
(JSC::Wasm::ExpressionType&gt;::finalizeEntrypoints):
(JSC::Wasm::ExpressionType&gt;::emitLoopTierUpCheck):
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::B3IRGenerator::emitLoopTierUpCheck):
(JSC::Wasm::B3IRGenerator::nextIndexInBuffer):
(JSC::Wasm::B3IRGenerator::loadFromScratchBuffer):
* Source/JavaScriptCore/wasm/WasmBBQPlan.cpp:
(JSC::Wasm::BBQPlan::work):
(JSC::Wasm::BBQPlan::initializeCallees):
* Source/JavaScriptCore/wasm/WasmCallee.h:
* Source/JavaScriptCore/wasm/WasmIRGeneratorHelpers.h:
(JSC::Wasm::buildEntryBufferForCatch):
(JSC::Wasm::buildEntryBufferForCatchSIMD):
(JSC::Wasm::buildEntryBufferForCatchNoSIMD):
(JSC::Wasm::emitCatchPrologueShared):
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::loadValuesIntoBuffer):
(JSC::Wasm::doOSREntry):
* Source/JavaScriptCore/wasm/WasmOperations.h:

Canonical link: <a href="https://commits.webkit.org/258282@main">https://commits.webkit.org/258282@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ce45dba01a43dc73f5bd0a9187603af5da0a1082

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/101442 "5 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/10600 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/34496 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/110712 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/170968 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/11552 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/1453 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/93829 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/108529 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/107224 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/8788 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/92028 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/35311 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/90686 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/23436 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/78311 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/91870 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/4202 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/24947 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/88016 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/1789 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/4264 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/1369 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29318 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/10350 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/44436 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/90916 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5694 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/6027 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/20307 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on Source/JavaScriptCore/wasm/WasmOperations.cpp:
```diff
@@ -109,49 +109,70 @@ static void triggerOMGReplacementCompile(TierUpCount& tierUp, OMGCallee* replace
 {
+    ASSERT(valueSize == 1 || valueSize == 2);
+    ASSERT(Options::useWebAssemblySIMD() || valueSize == 1);
+    bool usesSIMD = (valueSize == 2);
```
This could be swapped I guess? Also, it would be nice to have some kind of helper for valueSize, but I  am not sure what to call it.

justinmichaud commented on Source/JavaScriptCore/wasm/WasmOperations.cpp:
```diff
@@ -185,7 +208,7 @@ static void doOSREntry(Instance* instance, Probe::Context& context, BBQCallee& c
             context.gpr(entry.reg().gpr()) = *bitwise_cast<UCPURegister*>(bitwise_cast<uint8_t*>(context.fp()) + entry.offset());
         else
-            context.fpr(entry.reg().fpr()) = *bitwise_cast<double*>(bitwise_cast<uint8_t*>(context.fp()) + entry.offset());
+            (usesSIMD ? context.vector(entry.reg().fpr()).f64x2[1] : context.fpr(entry.reg().fpr())) = *bitwise_cast<double*>(bitwise_cast<uint8_t*>(context.fp()) + entry.offset());
```
Ignore this

Constellation commented on Source/JavaScriptCore/assembler/MacroAssemblerARMv7.cpp:
```diff
@@ -365,7 +365,7 @@ asm (
 #endif // COMPILER(GCC_COMPATIBLE)
 
-void MacroAssembler::probe(Probe::Function function, void* arg)
+void MacroAssembler::probe(Probe::Function function, void* arg, bool)
```
Since it is hard to know what this bool means, let's use enum.

Constellation commented on Source/JavaScriptCore/assembler/ProbeContext.h:
```diff
@@ -99,11 +116,11 @@ T CPUState::spr(SPRegisterID id) const
 }
 
-template<typename T>
+template<typename T, bool usesSIMD>
```
Since it is hard to know what this bool means, let's use enum.

Constellation commented on Source/JavaScriptCore/wasm/WasmCallee.h:
```diff
@@ -179,9 +179,14 @@ class OSREntryCallee final : public OptimizingJITCallee {
+    static Ref<BBQCallee> create(Wasm::Entrypoint&& entrypoint, size_t index, std::pair<const Name*, 
+        RefPtr<NameSection>>&& name, std::unique_ptr<TierUpCount>&& tierUpCount, Vector<UnlinkedWasmToWasmCall>&& unlinkedCalls,
+        StackMaps&& stackmaps, Vector<UnlinkedHandlerInfo>&& exceptionHandlers, Vector<CodeLocationLabel<ExceptionHandlerPtrTag>>&& exceptionHandlerLocations,
+        Vector<CodeLocationLabel<WasmEntryPtrTag>>&& loopEntrypoints, unsigned osrEntryScratchBufferSize, bool usesSIMD)
```
Since it is hard to know what this bool means, let's use enum.

Constellation commented on Source/JavaScriptCore/wasm/WasmCallee.h:
```diff
@@ -204,13 +209,19 @@ class BBQCallee final : public OptimizingJITCallee {
+    BBQCallee(Wasm::Entrypoint&& entrypoint, size_t index, std::pair<const Name*, RefPtr<NameSection>>&& name, std::unique_ptr<TierUpCount>&& tierUpCount, 
+        Vector<UnlinkedWasmToWasmCall>&& unlinkedCalls, StackMaps&& stackmaps, Vector<UnlinkedHandlerInfo>&& exceptionHandlers,
+        Vector<CodeLocationLabel<ExceptionHandlerPtrTag>>&& exceptionHandlerLocations, Vector<CodeLocationLabel<WasmEntryPtrTag>>&& loopEntrypoints, 
+        unsigned osrEntryScratchBufferSize, bool usesSIMD)
```
Since it is hard to know what this bool means, let's use enum.

Constellation reviewed (approved):
r=me

Commit:
SIMD values should be saved during OSR entry.
https://bugs.webkit.org/show_bug.cgi?id=249509

Reviewed by Yusuke Suzuki.

We update the masm probe, b3 and air generators to have a separate
osr entry path that saves the full value of vector registers.

A CPUState can now either have packed fprs or packed vectors, depending on
if the context's Air::Code has usesSIMD == true. This ensures that we
can still load/store pair in the common case, but we do waste some space.

The entry scratch buffer size is doubled when SIMD is used. While this is
wasteful, it is less likely to have bugs. Since this will only regress SIMD
functions, it is hopefully not too bad.

* JSTests/wasm/stress/simd-osr.js:
* Source/JavaScriptCore/assembler/MacroAssembler.h:
* Source/JavaScriptCore/assembler/MacroAssemblerARM64.cpp:
(JSC::MacroAssembler::probe):
* Source/JavaScriptCore/assembler/ProbeContext.h:
(JSC::Probe::CPUState::fpr):
(JSC::Probe::CPUState::vector):
(JSC::Probe::Context::vector):
* Source/JavaScriptCore/wasm/WasmAirIRGeneratorBase.h:
(JSC::Wasm::ExpressionType>::finalizeEntrypoints):
(JSC::Wasm::ExpressionType>::emitLoopTierUpCheck):
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::B3IRGenerator::emitLoopTierUpCheck):
(JSC::Wasm::B3IRGenerator::nextIndexInBuffer):
(JSC::Wasm::B3IRGenerator::loadFromScratchBuffer):
* Source/JavaScriptCore/wasm/WasmBBQPlan.cpp:
(JSC::Wasm::BBQPlan::work):
(JSC::Wasm::BBQPlan::initializeCallees):
* Source/JavaScriptCore/wasm/WasmCallee.h:
* Source/JavaScriptCore/wasm/WasmIRGeneratorHelpers.h:
(JSC::Wasm::buildEntryBufferForCatch):
(JSC::Wasm::buildEntryBufferForCatchSIMD):
(JSC::Wasm::buildEntryBufferForCatchNoSIMD):
(JSC::Wasm::emitCatchPrologueShared):
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::loadValuesIntoBuffer):
(JSC::Wasm::doOSREntry):
* Source/JavaScriptCore/wasm/WasmOperations.h:

Canonical link: https://commits.webkit.org/258282@main
